lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -962,6 +1001,83 @@ static LogicalResult buildAllocOp(ComponentLoweringState &componentState,
                     IntegerAttr::get(rewriter.getI1Type(), llvm::APInt(1, 1)));
   componentState.registerMemoryInterface(allocOp.getResult(),
                                          calyx::MemoryInterface(memoryOp));
+
+  unsigned elmTyBitWidth = memtype.getElementTypeBitWidth();
+  assert(elmTyBitWidth <= 64 && "element bitwidth should not exceed 64");
+  bool isFloat = !memtype.getElementType().isInteger();
+
+  auto shape = allocOp.getType().getShape();
+  int totalSize =
+      std::reduce(shape.begin(), shape.end(), 1, std::multiplies<int>());
+  // The `totalSize <= 1` check is a hack to:
+  // https://github.com/llvm/circt/pull/2661, where a multi-dimensional memory
+  // whose size in some dimension equals 1, e.g. memref<1x1x1x1xi32>, will be
+  // collapsed to `memref<1xi32>` with `totalSize == 1`. While the above case is
+  // a trivial fix, Calyx expects 1-dimensional memories in general:
+  // https://github.com/calyxir/calyx/issues/907
+  if (!(shape.size() <= 1 || totalSize <= 1)) {
+    allocOp.emitError("input memory dimension must be empty or one.");
+    return failure();
+  }
+
+  std::vector<uint64_t> flattenedVals(totalSize, 0);
+  if (isa<memref::GetGlobalOp>(allocOp)) {
+    auto getGlobalOp = cast<memref::GetGlobalOp>(allocOp);
+    auto *symbolTableOp =
+        getGlobalOp->template getParentWithTrait<mlir::OpTrait::SymbolTable>();
+    auto globalOp = dyn_cast_or_null<memref::GlobalOp>(
+        SymbolTable::lookupSymbolIn(symbolTableOp, getGlobalOp.getNameAttr()));
+    // Flatten the values in the attribute
+    auto cstAttr = llvm::dyn_cast_or_null<DenseElementsAttr>(
+        globalOp.getConstantInitValue());
+    int sizeCount = 0;
+    for (auto attr : cstAttr.template getValues<Attribute>()) {
+      assert((isa<mlir::FloatAttr, mlir::IntegerAttr>(attr)) &&
+             "memory attributes must be float or int");
+      if (auto fltAttr = dyn_cast<mlir::FloatAttr>(attr)) {
+        flattenedVals[sizeCount++] =
+            bit_cast<uint64_t>(fltAttr.getValueAsDouble());
+      } else {
+        auto intAttr = dyn_cast<mlir::IntegerAttr>(attr);
+        APInt value = intAttr.getValue();
+        flattenedVals[sizeCount++] = *value.getRawData();
+      }
+    }
+
+    rewriter.eraseOp(globalOp);
+  }
+
+  llvm::json::Array result;
+  result.reserve(std::max(static_cast<int>(shape.size()), 1));
+
+  bool isSignlessOrUnsigned = memtype.getElementType().isSignlessInteger() ||
+                              memtype.getElementType().isUnsignedInteger();
+  for (uint64_t bitValue : flattenedVals) {
+    llvm::json::Value value = 0;
+    if (isFloat) {
+      // we cast to `double` and let downstream calyx to deal with the actual
+      // value's precision handling
```

Comment:
Nit: Please use proper punctuation in your comments, i.e., capitalization and end with period.